compiler/rustc_transmute/src/maybe_transmutable.rs

```diff
@@ -0,0 +1,208 @@
+                self.answer_cached(src_state, dst_state_prime)
+            })
+        } else {
+            // for all `src_transition`s leading out of `src_state`,
+            for_all(
+                self.src_nfa.edges_from(src_state).unwrap_or(&empty_map),
+                |(src_transition, src_state_primes)| {
+                    // for all successive src states reached via `src_transition`
+                    for_all(src_state_primes, |&src_state_prime| {
+                        let dst_transitions =
+                            self.dst_nfa.edges_from(dst_state).unwrap_or(&empty_map);
+                        // there must exist at least one compatible dst_transition
+                        there_exists(dst_transitions, |(dst_transition, dst_state_primes)| {
+                            // that is compatible with the transition
+                            is_compatible(src_transition, dst_transition, self.within_references)
+                                .and(there_exists(dst_state_primes, |&dst_state_prime| {
+                                    // such that successive bytes of `src` are transmutable
+                                    // into some path of successive bytes of `dst`
+                                    self.answer_cached(src_state_prime, dst_state_prime)
+                                }))
+                        })
+                    })
+                },
+            )
+        }
+    }
+
+    #[inline(always)]
+    fn answer_cached(&mut self, src_state_prime: State, dst_state_prime: State) -> Answer<'tcx> {
+        if let Some(result) = self.cache.get(&(src_state_prime, dst_state_prime)) {
+            result.clone()
+        } else {
+            let result = self.answer(src_state_prime, dst_state_prime);
+            self.cache.insert((src_state_prime, dst_state_prime), result.clone());
+            result
+        }
+    }
+}
+
+pub fn is_compatible<'tcx>(
+    src_transition: &Transition<'tcx>,
+    dst_transition: &Transition<'tcx>,
+    within_references: bool,
+) -> Answer<'tcx> {
+    match (src_transition, dst_transition) {
+        (Transition::Byte(src_byte), Transition::Byte(dst_byte)) if src_byte == dst_byte => {
+            Answer::Yes
+        }
+        (Transition::Byte(_), Transition::Byte(Byte::UNINIT)) if !within_references => Answer::Yes,
+        (Transition::Ref(src), Transition::Ref(dst)) if src.min_align() >= dst.min_align() => {
+            Answer::IfTransmutable { src: src.clone(), dst: dst.clone() }
+        }
+        _ => Answer::No,
+    }
+}
+
+impl<'tcx> Answer<'tcx> {
+    fn and(self, rhs: Self) -> Self {
+        match (self, rhs) {
+            (Self::No, _) | (_, Self::No) => Self::No,
+            (Self::Yes, Self::Yes) => Self::Yes,
+            (Self::IfAll(mut lhs), Self::IfAll(ref mut rhs)) => {
+                lhs.append(rhs);
+                Self::IfAll(lhs)
+            }
+            (constraint, Self::IfAll(mut constraints))
```

Comment:
NB: this ends up reversing any short-circuiting that the other operations could achieve on `constraint`